### ISSUE TITLE
Feature Extraction Causing Problems

### ISSUE THREAD

**Issue by DroneMesh** [hill-a/stable-baselines#904]
Hi Everyone,

So I have noticed an issue when exporting due to the feature extraction. I am guessing that even if you are not using the image recognition part of the policy it still sets up those nodes in the graph. Now the issue arises when you want to export this to an ARM processor. tfcompile will automatically pickup that there is some sort conv network due to the nodes and require the arm compute library even though you do not neet it and thus making the hex much larger for no reason and in most cases the hex will not fit the arm mcu.

Any help into removing these and or way to turn that part off would be great thank you.

**Comment by Miffyli** [comment]
I am not sure why "feature extraction" would cause issues per-se, but I understand if convolutional layers would do this.

In any case, you might want to take a look at [exporting models documentation](https://stable-baselines.readthedocs.io/en/master/guide/export.html#manual-export). I recommend you only extract the network weights with `agent.get_parameters()`, and on your target machine re-create the network manually and then assign those parameters to correct places.

**Comment by DroneMesh** [comment]
@Miffyli Thank you for your quick response. I am not experienced enough to recreate the network manually at the moment but is there is a way to remove them. If you can you point me in the correct way and or help me. I would be able to create the docs for exporting into ARM devices and or C/C++

**Comment by Miffyli** [comment]
Unfortunately I am not familiar working with NNs in such platforms. However, if your network only consists of 1D input, 1D output and fully-connected layers ("MlpPolicy"), then you only need functions for matrix multiplication and any activation functions (tanh, in the case of stable-baselines), as well any processing needed for the output (e.g. softmax for discrete actions).  You can google for "Neural network in C++" or "MLP in C++" to find tutorials on this. Note that you do not need the "gradient" or "backpropagation" part (nor the derivatives), you can assume you already have the weights.

**Comment by DroneMesh** [comment]
@Miffyli  I am using PPO2, MLB, tanh, continuous action space. Can you point me to the source where the network gets initialized i was just looking at tensorboard and I think it's the flatten function at is invoking tfcompile to think this is a conv network?

**Comment by DroneMesh** [comment]
Ok so after digging some more I figured out the issue. There is no way to bypass the conv part of the policy initialization. The issue is arising from CNN and the MLP_Extracter in the policy.py so when tfcompile sees those parts in the graph it assumes its a conv network and asks for the libraries to be put into the compiled hex file and there is no space to add that library. 

Anyone know a workaround to not have them initials in the model. 

This makes this library unusable in most cases on small STM32 ARM processors which is very bad because this has great potential.

**Comment by Miffyli** [comment]
Again, I am not familiar with tfcompile and how it works, but you could try modifying the code by removing all references to convolutional networks (i.e. all imports of convolutional operations). If you only use "MlpPolicy", this should be rather easy. Other than that I do not have ideas to share.

**Comment by DroneMesh** [comment]
@Miffyli Exactly the case can you help me understand how to bypass that part i have been trying for paste 4 hours now. ANy suggestions would be very useful and it will also me to make a PR on exporting model.

**Comment by araffin** [comment]
I would suggest you to use SB3, the PyTorch version of SB, currrently in beta: https://github.com/DLR-RM/stable-baselines3
I have successfully installed it on a raspberry pi.

You can also compile the C++ lib only and easily export trained network to c++ with pytorch (cf pytorch doc).
In the documentation, we show also how to export Stable-Baselines2 models to pytorch: https://stable-baselines.readthedocs.io/en/master/guide/export.html

In the past, I also wrote some code to use trained cnn model in C++, using [eigen](https://github.com/sergionr2/RacingRobot/blob/12e33d305fb73ce37e19f8ec3b9f68fb0582c5b5/image_processing/c_extension/fast_image_processing.cpp) or [pytorch c api](https://github.com/sergionr2/RacingRobot/blob/12e33d305fb73ce37e19f8ec3b9f68fb0582c5b5/image_processing/pytorch_extension/custom_cnn.cpp).
However, I was still using pybind11 to pass python objects to the c++ code.

**Comment by DroneMesh** [comment]
@Miffyli I would love too however I am constrained with the TF version and AOT tfcompile. If i change my setup it will set me back a couple of months.

**Comment by DroneMesh** [comment]
So if i override the policy with a custom one like this it should work, correct me if I'm wrong.

I found this custom policy as a template but i would need to remove the flatten part and should in theory not build any sort image processing node within the network.


`class KerasPolicy(ActorCriticPolicy):
    def __init__(self, sess, ob_space, ac_space, n_env, n_steps, n_batch, reuse=False, **kwargs):
        super(KerasPolicy, self).__init__(sess, ob_space, ac_space, n_env, n_steps, n_batch, reuse=reuse, scale=False)

        with tf.variable_scope("model", reuse=reuse):
            flat = tf.keras.layers.Flatten()(self.processed_obs)
            x = tf.keras.layers.Dense(64, activation="tanh", name='pi_fc_0')(flat)
            pi_latent = tf.keras.layers.Dense(64, activation="tanh", name='pi_fc_1')(x)

            x1 = tf.keras.layers.Dense(64, activation="tanh", name='vf_fc_0')(flat)
            vf_latent = tf.keras.layers.Dense(64, activation="tanh", name='vf_fc_1')(x1)

            value_fn = tf.keras.layers.Dense(1, name='vf')(vf_latent)

            self._proba_distribution, self._policy, self.q_value = \
                self.pdtype.proba_distribution_from_latent(pi_latent, vf_latent, init_scale=0.01)

        self._value_fn = value_fn
        self._setup_init()

    def step(self, obs, state=None, mask=None, deterministic=False):
        if deterministic:
            action, value, neglogp = self.sess.run([self.deterministic_action, self.value_flat, self.neglogp],
                                                   {self.obs_ph: obs})
        else:
            action, value, neglogp = self.sess.run([self.action, self.value_flat, self.neglogp],
                                                   {self.obs_ph: obs})
        return action, value, self.initial_state, neglogp

    def proba_step(self, obs, state=None, mask=None):
        return self.sess.run(self.policy_proba, {self.obs_ph: obs})

    def value(self, obs, state=None, mask=None):
        return self.sess.run(self.value_flat, {self.obs_ph: obs})`

**Comment by Miffyli** [comment]
@DroneMesh Again, I do not know what is required to make this work, but to you should be able to get rid of all CNN stuff by removing references in [policies](https://github.com/hill-a/stable-baselines/blob/master/stable_baselines/common/policies.py#L538) (leave only MlpPolicy and remove all "cnn" stuff/set to None. Also remove nature_cnn) and remove [`conv` stuff in tf_layers](https://github.com/hill-a/stable-baselines/blob/master/stable_baselines/common/tf_layers.py). That should be all cnn-related operations. Not sure why these would be included in the compiling if they are not used, but I am not familiar with tfcompile.

**Comment by DroneMesh** [comment]
@Miffyli  I ended up removing everything related to CNN stuff however it turns out the issue is with the tensor values and nodes. tfcompile thinks there are long long int in the graph so I thought maybe the tensors are np.float64 but that was not the case. So i am digging some more into it now.

**Comment by DroneMesh** [comment]
Solved it was an LLVM bug
